src/charm.py

```diff
@@ -305,180 +142,70 @@ def _on_relation_departed(self, event: RelationDepartedEvent, charm_state: Charm
-        """
-        peer_relation = self.model.relations[synapse.SYNAPSE_PEER_RELATION_NAME]
-        if not peer_relation:
-            logger.error(
-                "Failed to get main unit: no peer relation %s found",
-                synapse.SYNAPSE_PEER_RELATION_NAME,
-            )
-            return None
-        return peer_relation[0].data[self.app].get(MAIN_UNIT_ID)
-
-    def get_main_unit_address(self) -> str:
-        """Get main unit address. If main unit is None, use unit name.
-
-        Returns:
-            main unit address as unit-0.synapse-endpoints.
-        """
-        main_unit_name = self.get_main_unit()
-        if main_unit_name is None:
-            main_unit_name = self.unit.name
-        main_unit_formatted = main_unit_name.replace("/", "-")
-        return f"{main_unit_formatted}.{self.app.name}-endpoints"
-
-    def set_main_unit(self, unit: str) -> None:
-        """Create/Renew an admin access token and put it in the peer relation.
-
-        Args:
-            unit: Unit to be the main.
-        """
-        peer_relation = self.model.relations[synapse.SYNAPSE_PEER_RELATION_NAME]
-        if not peer_relation:
-            logger.error(
-                "Failed to get main unit: no peer relation %s found",
-                synapse.SYNAPSE_PEER_RELATION_NAME,
-            )
-        else:
-            logging.info("Setting main unit to be %s", unit)
-            peer_relation[0].data[self.app].update({MAIN_UNIT_ID: unit})
+    def reconcile(self, charm_state: CharmState) -> None:
+        """Reconcile Synapse configuration with charm state.
 
-    def set_signing_key(self, signing_key: str) -> None:
-        """Create secret with signing key content.
+        This is the main entry for changes that require a restart.
 
         Args:
-            signing_key: signing key as string.
+            charm_state: Instance of CharmState
         """
-        peer_relation = self.model.relations[synapse.SYNAPSE_PEER_RELATION_NAME]
-        if not peer_relation:
-            logger.error(
-                "Failed to set signing key: no peer relation %s found",
-                synapse.SYNAPSE_PEER_RELATION_NAME,
-            )
+        container = self.unit.get_container(synapse.SYNAPSE_CONTAINER_NAME)
+        if not container.can_connect():
+            self.unit.status = ops.MaintenanceStatus("Waiting for Synapse pebble")
             return
-
-        if signing_key == self.get_signing_key():
-            logger.info("Received signing key but there is no change, skipping")
+        if helpers.is_redis_required(self, charm_state):
+            self.unit.status = ops.BlockedStatus("Redis integration is required.")
             return
+        self.model.unit.status = ops.MaintenanceStatus("Configuring Synapse")
+        helpers.configure_and_start_services(self, charm_state, container)
+        if helpers.is_mjolnir_enabled(self, charm_state):
+            self._mjolnir.enable(charm_state)
```

Comment:
Shouldn't we check if mjolnir is started (similarly to the nginx / synapse check) ?